pandas/core/indexes/base.py

```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
 
     def _join_multi(self, other, how, return_indexers=True):
         from .multi import MultiIndex
-        self_is_mi = isinstance(self, MultiIndex)
-        other_is_mi = isinstance(other, MultiIndex)
 
         # figure out join names
         self_names = [n for n in self.names if n is not None]
         other_names = [n for n in other.names if n is not None]
-        overlap = list(set(self_names) & set(other_names))
+        overlap = set(self_names) & set(other_names)
 
-        # need at least 1 in common, but not more than 1
+        # need at least 1 in common
         if not len(overlap):
-            raise ValueError("cannot join with no level specified and no "
-                             "overlapping names")
-        if len(overlap) > 1:
-            raise NotImplementedError("merging with more than one level "
-                                      "overlap on a multi-index is not "
-                                      "implemented")
-        jl = overlap[0]
+            raise ValueError("cannot join with no overlapping index names")
+
+        # Drop the non matching levels
+        ldrop_levels = list(set(self_names) - overlap)
+        rdrop_levels = list(set(other_names) - overlap)
+
+        self_is_mi = isinstance(self, MultiIndex)
+        other_is_mi = isinstance(other, MultiIndex)
+
+        if self_is_mi and other_is_mi:
+            self_tmp = self.droplevel(ldrop_levels)
+            other_tmp = other.droplevel(rdrop_levels)
+
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise ValueError("Join on level between two MultiIndex objects"
+                                 "is ambiguous")
+
+            # Append to the returned Index the non-overlapping levels
+            dropped_lvls = ldrop_levels + rdrop_levels
+
+            join_index, lidx, ridx = self_tmp.join(other_tmp, how=how,
+                                                   return_indexers=True)
+
+            if how == 'outer':
+                levels, labels, names = self._complete_outer_join(other,
+                                                                  join_index,
+                                                                  lidx, ridx,
+                                                                  dropped_lvls)
+
+            else:
+                if how == 'left':
+                    join_index = self
+                elif how == 'right':
+                    join_index = other
+
+                levels = join_index.levels
```

Comment:
move these 3 assignments after the .join, then you can make the 'how' an if/elif (for each of outer/left/right)